File: array/172.cpp
#include <bits/stdc++.h>

using namespace std;

using vi = vector<int>;
using vvi = vector<vector<int>>;
int findShortestSubArray(vector<int>& nums) {
    vvi ans(50000);
    int cont = 0;
    int distance = INT_MAX;
    int start, end;

    for(int i = 0; i < nums.size(); ++i) {
        ans[nums[i]].push_back(i);
        int freq = ans[nums[i]].size();
        start = *ans[nums[i]].begin();
        end = *(--ans[nums[i]].end());

        if(freq > cont) {
            distance = end - start;
            cont = freq;
        }
        else if (freq == cont) {
            distance = (end - start) < distance? end - start: distance;
        }
    }

    return distance + 1;
}

int main() {
    vi nums = {1,2,2,3,1, 49999,49999, 49999};
    cout<<findShortestSubArray(nums);

    return 0;
}

File: array/172.py
from typing import List


def findShortestSubArray(nums: List[int]) -> int:
    ans = {}
    distance = 2**31 - 1
    cont = 0

    for index, value in enumerate(nums):
        if not value in ans:
            ans[value] = [index, index, 1]
        else:
            ans[value][2] += 1
            ans[value][0] = index

        if ans[value][2] > cont:
            distance = ans[value][0] - ans[value][1]
            cont = ans[value][2]
        elif ans[value][2] == cont:
            diff = ans[value][0] - ans[value][1]
            if diff < distance:
                distance = diff

    return distance + 1


print(findShortestSubArray([1, 2, 2, 3, 1]))
